app/models/katello/glue/provider.rb

```diff
@@ -144,14 +150,39 @@ def import_product(product_json)
         if product&.redhat?
           product.update!(:name => product_json['name']) unless product.name == product_json['name']
         elsif product.nil?
-          product = Glue::Candlepin::Product.import_from_cp(product_json, organization)
+          # we have an orphaned Red Hat product
+          product = Glue::Candlepin::Product.import_redhat_product_from_cp(product_json, organization)
         end
         product
       end
 
-      def index_subscriptions
-        Katello::Subscription.import_all(organization)
-        Katello::Pool.import_all(organization, false)
+      def orphaned_custom_product?(product_json, organization)
+        return unless orphaned_product?(product_json, organization)
+        product_provider = provider_for_cp_id(cp_id: product_json['id'], organization: organization)
+        unless product_provider.redhat_provider?
+          Rails.logger.info "Found orphaned custom product #{product_json['id']} in Candlepin: '#{product_json['name']}'. Skipping import into Katello; run rake katello:delete_orphaned_custom_products to remove it from Candlepin."
+          true
+        else
+          false
+        end
+      end
+
+      def orphaned_product?(product_json, organization)
+        organization.products.find_by(:cp_id => product_json['id']).blank?
+      end
+
+      def provider_for_cp_id(cp_id:, organization:)
+        return organization.redhat_provider unless ::Katello::Glue::Candlepin::Product.engineering_product_id?(cp_id)
+        if ::Katello::Glue::Candlepin::Product.custom_product_id?(cp_id)
+          organization.anonymous_provider
+        else
+          organization.redhat_provider
+        end
+      end
+
+      def index_subscriptions(skip_cp_ids: [])
+        Katello::Subscription.import_all(organization, nil, skip_cp_ids)
```

Comment:
Would it make sense for `import_all` to have a param called `skip_orphaned` that then does the thing instead of passing them down through `import_all` ?

Reply:
not sure I understand this suggestion.?

Reply:
It really was just a question! Worded differently: whose responsibility is it to know what constitutes an orphaned (or skipped) product? Should the import_all implementation care about determining that  (rather than being told those values), or is this the best place for it - the answer might be yes :) I'm a little fuzzy on the whole execution path here so I can't say

Reply:
ohh, gotcha! I did play around with this a bit, and it does clean up the code quite a bit not to have to pass around `skip_product_ids` or `skip_cp_ids` anywhere. In addition, I discovered that `Subscription.import_all` is also called in a previous upgrade rake task, fix_invalid_pools.rake, so we must account for that too. This made me realize that whenever we call `import_all` in the future, it _will_ break if you have orphaned CP products and don't pass them in. So good catch!

But making that change introduces two complications: (a) now, in addition to `import_all`, `import_products` also needs to know which ones to skip; and (b) in the case of `import_all`, since it's for Pools and Subscriptions the field to look at is `object['productId']` whereas for Products it's `object['id']`. This makes it either impossible or really difficult to keep that knowledge in `import_all`. For this reason I think it's better to keep it as is, but make it a required keyword arg. I'll push shortly with what I mean. Thoughts?

Reply:
for the record, I'm still not sure about this. Even with this new way, if you pass `[]` just to satisfy the required argument, but you actually have orphaned CP products for which you're reimporting their pools, it will still break. Still need to think some more.

Reply:
What's the use case for calling import_all without omitting the orphans? Having a clear answer to that might help here?

Reply:
The only use case I can think of for calling `import_all` without passing in the orphans is if you don't know about them. If I'm in the process of reimporting products, and I come across an orphaned product, then I know that its associated pools are also orphaned, so I can skip them on import as well. But if I'm only importing pools or subscriptions (like in fix_invalid_pools.rake), I won't know what's valid or not, and my only choice is to try to reimport everything.  :thinking:

Reply:
I see what you're saying. It feels like there's a missing layer of abstraction such as a service class that encapsulates the logic around all of that (which itself would call `import_all` for one or more candlepin-backed models) - cause it seems to be spread out and somewhat unclear. Not that you should attempt to solve that here, but not having it forces you to pay attention to changing code in more areas of the app.